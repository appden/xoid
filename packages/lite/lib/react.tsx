import { useReducer, useEffect, useLayoutEffect, useRef } from 'react'
import { create, subscribe, select } from './vanilla'
import { Observable } from '@xoid/engine'

// For server-side rendering: https://github.com/react-spring/zustand/pull/34
const useIsoLayoutEffect = typeof window === 'undefined' ? useEffect : useLayoutEffect

const useConstant = <T extends any>(fn: () => T): T => {
  const ref = useRef<{ c: T }>()
  if (!ref.current) ref.current = { c: fn() }
  return ref.current.c
}

/**
 * Subscribes to a store, or a value inside a React function component.
 * @see [xoid.dev/docs/api-react/usestore](https://xoid.dev/docs/api-react/usestore)
 */

export function useStore<T>(store: Observable<T>): T
export function useStore<T, U>(store: Observable<T>, selector: (state: T) => U): U
export function useStore<T, U>(store: Observable<T>, selector?: (state: T) => U): any {
  const forceUpdate = useReducer((c) => c + 1, 0)[1]
  const item = useConstant(() => {
    return selector ? select(store, selector) : store
  })

  useIsoLayoutEffect(() => subscribe(item, forceUpdate), [])
  return item()
}

/**
 * Subscribes to a store, or a value inside a React function component.
 * @see [xoid.dev/docs/api-react/usesetup](https://xoid.dev/docs/api-react/usesetup)
 */

export function useSetup<T>(
  model: (deps: Observable<undefined>, onCleanup: (fn: () => void) => void) => T
): T
export function useSetup<T, P>(
  model: (deps: Observable<P>, onCleanup: (fn: () => void) => void) => T,
  props: P
): T
export function useSetup(model: (deps: any, onCleanup: any) => any, props?: any): any {
  const setup = useConstant(() => {
    const deps = create(props)
    const fns: any[] = []
    const onCleanup = (fn: any) => fns.push(fn)
    const main = model(deps, onCleanup)
    return { main, deps, fns }
  })
  // eslint-disable-next-line react-hooks/exhaustive-deps
  useIsoLayoutEffect(() => setup.deps(props), [props])
  // eslint-disable-next-line react-hooks/exhaustive-deps
  useIsoLayoutEffect(() => () => setup.fns.forEach((fn) => fn()), [])
  return setup.main
}
