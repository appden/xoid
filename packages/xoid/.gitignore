types
index.js
index.esm.js