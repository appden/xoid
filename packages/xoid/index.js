'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var C = require('@xoid/core');
var M = require('@xoid/model');
var R = require('@xoid/ready');
var x = Object.assign(M.model.bind({}), C, M, R);

function make(O) {
	Object.keys(O).forEach(function (k) {
		Object.defineProperty(exports, k, {
			enumerable: true,
			get: function () {
				return O[k];
			}
		});
	});
}
make(C)
make(M)
make(R)
exports.default = x;