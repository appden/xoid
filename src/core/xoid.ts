import { List, Store, ReverseTransform, Actor } from './types'
import { deepClone, memberMap, parentMap, storeMap } from './utils'

export const get = <T extends object>(item: T): ReverseTransform<T> => {
  const record = storeMap.get(item) || memberMap.get(item)
  if (record) return record.value
  else throw TypeError('TODO: cannot get non-observable value')
}

export const set = <T extends object>(
  item: T,
  fn:
    | ReverseTransform<T>
    | ((state: ReverseTransform<T>) => ReverseTransform<T>)
): void => {
  const record = storeMap.get(item) || memberMap.get(item)
  if (record) {
    const { value, address, internal } = record
    const newValue = typeof fn === 'function' ? (fn as Function)(value) : fn
    if (newValue !== value) {
      if (address.length) {
        const newState = { ...internal.get() }
        address.reduce((acc0: any, key: any, i: number) => {
          if (i === address.length - 1) acc0[key] = newValue
          return acc0[key]
        }, newState)
        internal.set(newState)
      } else {
        internal.set(newValue)
      }
    }
  } else throw TypeError('TODO: cannot get non-observable value')
}

export const use = <T, A>(item: Store<T, A>): A => {
  const record = storeMap.get(item)
  if (record) return record.internal.getActions()
  else throw TypeError('TODO: cannot use non-store')
}

export const destroy = <T extends Store<any, any>>(item: T) => {
  const record = storeMap.get(item as object)
  if (record) return record.internal.destroy()
  else throw TypeError('TODO: cannot destroy non-store')
}

export const subscribe = <T extends Store<any, any>>(
  item: T,
  fn: (state: ReverseTransform<T>) => void
) => {
  const record = storeMap.get(item as object) || memberMap.get(item)
  if (record) {
    const { address, internal } = record
    if (address.length) {
      return internal.subscribe(fn as any, (state: any) =>
        address.reduce((acc: any, key: any) => {
          return acc[key]
        }, state)
      )
    } else {
      return internal.subscribe(fn as any)
    }
  } else {
    throw TypeError('TODO: cannot subscribe non-observable')
  }
}

export const parent = <T extends Store<any>>(item: T): Store<any> => {
  const record = parentMap.get(item as object)
  if (record) return record.parent
}
