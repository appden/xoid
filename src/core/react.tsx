import { useReducer, useMemo, useEffect, useLayoutEffect, useRef } from 'react'
import { META } from './createCell'
import { resolverFactory } from './resolverFactory'
import { DetransformInner, Unsub, Value } from './types'
import { watchHelper } from './utils'

// For server-side rendering: https://github.com/react-spring/zustand/pull/34
const useIsoLayoutEffect =
  typeof window === 'undefined' ? useEffect : useLayoutEffect

/**
 * Subscribes to a store, or a value inside a React function component.
 * @see [xoid.dev/docs/api/use-store](https://xoid.dev/docs/api/use-store)
 */

export function useStore<T>(store: Value<T>): DetransformInner<T> {
  const forceUpdate = useReducer((c) => c + 1, 0)[1]
  const [unsubAll, getter] = useMemo(() => watchHelper(store, forceUpdate), [
    store,
  ])
  useIsoLayoutEffect(() => unsubAll)
  return getter()
}

const useConstant = <T extends any>(fn: () => T): T => {
  const ref = useRef<{ c: T }>()
  if (!ref.current) ref.current = { c: fn() }
  return ref.current.c
}
