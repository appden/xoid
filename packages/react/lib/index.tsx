import { useReducer, useEffect, useLayoutEffect, useRef, useMemo } from 'react'
// @ts-ignore
import { Atom, subscribe, create, select } from 'xoid'

// For server-side rendering: https://github.com/react-spring/zustand/pull/34
const useIsoLayoutEffect = typeof window === 'undefined' ? useEffect : useLayoutEffect

const useConstant = <T extends any>(fn: () => T): T => {
  const ref = useRef<{ c: T }>()
  if (!ref.current) ref.current = { c: fn() }
  return ref.current.c
}

/**
 * Subscribes to an xoid atom inside a React function component.
 * @see [xoid.dev/docs/api-react/usestore](https://xoid.dev/docs/api-react/usestore)
 */

export function useAtom<T>(store: Atom<T>): T
export function useAtom<T, U>(store: Atom<T>, selector: (state: T) => U): U
export function useAtom<T, U extends keyof T>(store: Atom<T>, selector: U): T[U]
export function useAtom<T, U>(store: Atom<T>, selector?: (state: T) => U): any {
  const forceUpdate = useReducer((c) => c + 1, 0)[1]
  const item = useMemo(() => {
    return selector ? select(store, selector) : store
  }, [store, selector])

  useIsoLayoutEffect(() => subscribe(item, forceUpdate), [])
  return item()
}

/**
 * Can be used to create local state inside React components. Similar to `React.useMemo`.
 * @see [xoid.dev/docs/api-react/usesetup](https://xoid.dev/docs/api-react/usesetup)
 */

export function useSetup<T>(
  model: (deps: Atom<undefined>, onCleanup: (fn: () => void) => void) => T
): T
export function useSetup<T, P>(
  model: (deps: Atom<P>, onCleanup: (fn: () => void) => void) => T,
  props: P
): T
export function useSetup(model: (deps: any, onCleanup: any) => any, props?: any): any {
  const setup = useConstant(() => {
    const deps = create(props)
    const fns: any[] = []
    const onCleanup = (fn: any) => fns.push(fn)
    const main = model(deps, onCleanup)
    return { main, deps, fns }
  })
  // eslint-disable-next-line react-hooks/exhaustive-deps
  useIsoLayoutEffect(() => setup.deps(props), [props])
  // eslint-disable-next-line react-hooks/exhaustive-deps
  useIsoLayoutEffect(() => () => setup.fns.forEach((fn) => fn()), [])
  return setup.main
}
