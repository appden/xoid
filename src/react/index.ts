import { useReducer, useEffect, useLayoutEffect } from 'react'
import { get, set, subscribe, use } from '../core'
import { Decorator, StateOf, Store, Value } from '../core/types'
import { getData, isRootData } from '../core/utils'

// For server-side rendering: https://github.com/react-spring/zustand/pull/34
const useIsoLayoutEffect =
  typeof window === 'undefined' ? useEffect : useLayoutEffect

export type SetState<T> = (
  state: StateOf<T> | ((state: StateOf<T>) => StateOf<T> | Promise<StateOf<T>>),
  decorator?: Decorator<T>
) => void

export function useStore<S extends Value<any>>(
  store: S
): S extends Value<infer T> | Store<infer T>
  ? [StateOf<T>, SetState<T>]
  : S extends Store<infer T, infer A>
  ? [StateOf<T>, A]
  : never {
  const data = getData(store)
  if (!data)
    throw TypeError(
      '[ @xoid/react ]: Argument of `useStore` should be a Store or a Store member.'
    )
  const isStore = isRootData(data)
  const [, forceUpdate] = useReducer((c) => c + 1, 0) as [never, () => void]
  useIsoLayoutEffect(() => subscribe(store, forceUpdate), [store])
  const setState = (value: any) => set(store, value)
  //@ts-ignore
  return [get(store), isStore ? use(store) || setState : setState]
}
