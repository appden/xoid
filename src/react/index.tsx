import React, {
  useReducer,
  useMemo,
  useEffect,
  useLayoutEffect,
  useRef,
} from 'react'
import { create } from '../core'
import { Store, DetransformInner, Value } from '../core/types'
import { watchHelper } from '../core/utils'

// For server-side rendering: https://github.com/react-spring/zustand/pull/34
const useIsoLayoutEffect =
  typeof window === 'undefined' ? useEffect : useLayoutEffect

const useConstant = <T extends any>(fn: () => T): T => {
  const ref = useRef<{ c: T }>()
  if (!ref.current) ref.current = { c: fn() }
  return ref.current.c
}

/**
 * Subscribes to a store, or a value inside a React function component.
 * @see [xoid.dev/docs/api/use-store](https://xoid.dev/docs/api/use-store)
 */

export function useStore<T>(store: Value<T>): DetransformInner<T> {
  const forceUpdate = useReducer((c) => c + 1, 0)[1]
  const [unsubAll, getter] = useMemo(() => watchHelper(store, forceUpdate), [
    store,
    forceUpdate,
  ])
  useIsoLayoutEffect(() => unsubAll)
  return getter()
}

/**
 * Subscribes to a store, or a value inside a React function component.
 * @see [xoid.dev/docs/api/use-local](https://xoid.dev/docs/api/use-local)
 */

export function useLocal<S>(model: () => S): any
export function useLocal<S, P>(model: (deps: Store<P>) => S, props: P): any
export function useLocal<S, P>(
  model: (deps?: Store<P | undefined>) => S,
  props?: P
): any {
  const deps = useConstant(() => create(props))
  const isMounted = useConstant(() => create(false))
  // eslint-disable-next-line react-hooks/exhaustive-deps
  useEffect(() => {
    const im = isMounted()
    if (im) deps(props)
    else isMounted(true)
  }, [props])
  return useConstant(() => model(deps))
}

/**
 * Subscribes to a store, or a value inside a React function component.
 * @see [xoid.dev/docs/api/reactive](https://xoid.dev/docs/api/reactive)
 */

const x = (store: Store<string | number>) => {
  return <Xoid store={store} />
}

const Xoid = (props: any) => {
  const state = useStore(props.store)
  return <>{state}</>
}

export default x
