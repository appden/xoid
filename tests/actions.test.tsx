import { render } from '@testing-library/react'
import { useAtom } from '@xoid/react'
import React from 'react'
import { create } from 'xoid'
import { debug } from './testHelpers'

it('uses the actions in vanilla (interop)', async () => {
  const atom = create({ count: 0 }, (atom) => ({
    inc: () => atom.update((state) => ({ count: state.count + 1 })),
  }))
  atom.actions.inc()
  expect(debug(atom)).toMatchSnapshot()
})

it('uses the actions in vanilla', async () => {
  const atom = create({ count: 0 }, (atom) => ({
    inc: () => atom.update((state) => ({ count: state.count + 1 })),
  }))
  atom.actions.inc()
  expect(debug(atom)).toMatchSnapshot()
})

it('uses the actions in React', async () => {
  const atom = create({ count: 0 }, (atom) => ({
    inc: () => atom.update((state) => ({ count: state.count + 1 })),
  }))

  function Counter() {
    const { count } = useAtom(atom)
    const { inc } = atom.actions
    React.useEffect(inc, [inc])
    return <div>count: {count}</div>
  }

  const { findByText } = render(<Counter />)

  await findByText('count: 1')
})
