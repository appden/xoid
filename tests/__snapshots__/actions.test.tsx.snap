// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`uses the actions in vanilla 1`] = `
Object {
  "get": Object {
    "count": 1,
  },
  "getSerialized": "{\\"count\\":1}",
  "self": Object {
    "focus": [Function],
    "map": [Function],
    "set": [Function],
    "subscribe": [Function],
    "update": [Function],
    "value": Object {
      "count": 1,
    },
    "watch": [Function],
    Symbol(): Object {
      "atom": [Circular],
      "get": [Function],
      "isStream": false,
      "listeners": Set {},
      "set": [Function],
      "subscribe": [Function],
      "usable": Object {
        "inc": [Function],
      },
    },
  },
  "selfSerialized": "{\\"value\\":{\\"count\\":1}}",
  "use": Object {
    "inc": [Function],
  },
}
`;
